src/transaction/transaction_single_node_group_complete_manager.cpp

```diff
@@ -0,0 +1,309 @@
+  {
+    /* This function is called after adding a transaction to the current group. */
+    assert (get_current_group ().get_container ().size () >= 1);
+#if defined (SERVER_MODE)
+    if (is_latest_closed_group_completed ())
+      {
+	/* We can move the next calls outside of mutex. */
+	if (can_wakeup_group_complete_daemon (true)
+	    || pgbuf_has_perm_pages_fixed (&cubthread::get_entry ()))
+	  {
+	    /* This means that GC thread didn't start yet group close. */
+	    gl_single_node_gcm_daemon->wakeup ();
+	  }
+      }
+    else if (!is_latest_closed_group_complete_started ()
+	     && is_latest_closed_group_prepared_for_complete ())
+      {
+	/* Be sure that log flush knows that GC thread waits for it. */
+	log_wakeup_log_flush_daemon ();
+      }
+#endif
+  }
+
+#if defined (SERVER_MODE)
+  //
+  // can_wakeup_group_complete_daemon - true, if can wakeup group complete daemon
+  //
+  bool single_node_group_complete_manager::can_wakeup_group_complete_daemon (bool inc_gc_request_count)
+  {
+    bool can_wakeup_GC;
+
+    if (!LOG_IS_GROUP_COMMIT_ACTIVE ())
+      {
+	/* non-group commit */
+	can_wakeup_GC = true;
+      }
+    else
+      {
+	/* group commit */
+	can_wakeup_GC = false;
+
+	if (inc_gc_request_count)
+	  {
+	    log_Stat.gc_commit_request_count++;
+	  }
+      }
+
+    return can_wakeup_GC;
+  }
+#endif
+
+  //
+  // can_close_current_group check whether the current group can be closed.
+  //
+  bool single_node_group_complete_manager::can_close_current_group ()
+  {
+    if (!is_latest_closed_group_completed ())
+      {
+	/* Can't advance to the next group since the current group was not completed yet. */
+	return false;
+      }
+
+    if (is_current_group_empty ())
+      {
+	// no transaction, can't close the group.
+	return false;
+      }
+
+    return true;
+  }
+
+  //
+  // prepare_complete prepares group complete. Always should be called before do_complete.
+  //
+  void single_node_group_complete_manager::do_prepare_complete (THREAD_ENTRY *thread_p)
+  {
+    LOG_LSA closed_group_start_complete_lsa, closed_group_end_complete_lsa;
+    LOG_TDES *tdes;
+    bool has_postpone;
+
+    if (close_current_group ())
+      {
+	cubstream::stream_position closed_group_stream_start_position = 0, closed_group_stream_end_position = 0;
+	tx_group &closed_group = get_latest_closed_group ();
+	tdes = logtb_get_tdes (thread_p);
+
+	/* TODO - Introduce parameter. For now complete group MVCC only here. Notify MVCC complete. */
+	log_Gl.mvcc_table.complete_group_mvcc (thread_p, closed_group);
+	notify_group_mvcc_complete (closed_group);
+
+	if (!HA_DISABLED () && css_ha_server_state () == HA_SERVER_STATE_ACTIVE)
+	  {
+	    /* This is a single node that must generate stream group commits. */
+	    tdes->get_replication_generator ().pack_group_commit_entry (closed_group_stream_start_position,
+		closed_group_stream_end_position);
+	  }
+
+	log_append_group_complete (thread_p, tdes, 0,
```

Comment:
stream position is needed for recovery m_active_start_position

Reply:
Indeed, for master without slave case.